jupyterhub/app.py

```diff
@@ -1975,25 +1975,63 @@ async def init_groups(self):
                 group.users.append(user)
         db.commit()
 
-    async def init_roles(self):
+    async def init_role_creation(self):
         """Load default and predefined roles into the database"""
-        db = self.db
-        # tokens are added separately
-        role_bearers = ['users', 'services', 'groups']
-
-        # load default roles
         self.log.debug('Loading default roles to database')
         default_roles = roles.get_default_roles()
-        for role in default_roles:
-            roles.create_role(db, role)
+        config_role_names = [r['name'] for r in self.load_roles]
+        init_roles = self.load_roles.copy()
+        for role_name in config_role_names:
+            if config_role_names.count(role_name) > 1:
+                raise AttributeError(
+                    f"Role {role_name} multiply defined. Please check the `load_roles` configuration"
+                )
+        for role_spec in default_roles:
+            if role_spec['name'] not in config_role_names:
+                init_roles.append(role_spec)
+        if not orm.Role.find(self.db, name='admin'):
+            self._rbac_upgrade = True
+        self.db.query(orm.Role).delete()
```

Comment:
I don't think we want to delete all the roles on every startup. This will make startup very expensive with lots of users, as we'll need to reassign potentially thousands of users to the `user` role on each launch.

Instead, just delete the roles whose name is not in config_role_names. Then it will be up to the role assignment step to ensure everybody gets the right assignments (and unassignments).

Deleting here will also unassign token roles, which won't get reassigned later (another case to test!).

Reply:
Did not consider that case but indeed, I'll only overwrite the roles that are not present in current config